src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs

```diff
@@ -49,20 +49,48 @@ private static unsafe int WaitForMultipleObjectsIgnoringSyncContext(IntPtr* pHan
             Thread currentThread = Thread.CurrentThread;
             currentThread.SetWaitSleepJoinState();
 
-#if NATIVEAOT
-            int result;
-            if (reentrantWait)
+            ulong startTime = 0;
+            if (millisecondsTimeout != -1)
             {
-                Debug.Assert(!waitAll);
-                result = RuntimeImports.RhCompatibleReentrantWaitAny(false, millisecondsTimeout, numHandles, pHandles);
+                startTime = Interop.Kernel32.GetTickCount64();
             }
-            else
+
+            int result;
+            while (true)
             {
-                result = (int)Interop.Kernel32.WaitForMultipleObjectsEx((uint)numHandles, (IntPtr)pHandles, waitAll ? Interop.BOOL.TRUE : Interop.BOOL.FALSE, (uint)millisecondsTimeout, Interop.BOOL.FALSE);
-            }
+#if NATIVEAOT
+                if (reentrantWait)
+                {
+                    Debug.Assert(!waitAll);
+                    result = RuntimeImports.RhCompatibleReentrantWaitAny(true, millisecondsTimeout, numHandles, pHandles);
+                }
+                else
+                {
+                    result = (int)Interop.Kernel32.WaitForMultipleObjectsEx((uint)numHandles, (IntPtr)pHandles, waitAll ? Interop.BOOL.TRUE : Interop.BOOL.FALSE, (uint)millisecondsTimeout, Interop.BOOL.TRUE);
+                }
 #else
-            int result = (int)Interop.Kernel32.WaitForMultipleObjectsEx((uint)numHandles, (IntPtr)pHandles, waitAll ? Interop.BOOL.TRUE : Interop.BOOL.FALSE, (uint)millisecondsTimeout, Interop.BOOL.FALSE);
+                result = (int)Interop.Kernel32.WaitForMultipleObjectsEx((uint)numHandles, (IntPtr)pHandles, waitAll ? Interop.BOOL.TRUE : Interop.BOOL.FALSE, (uint)millisecondsTimeout, Interop.BOOL.TRUE);
 #endif
+
+                if (result != Interop.Kernel32.WAIT_IO_COMPLETION)
+                    break;
+
+                // Handle APC completion by adjusting timeout and retrying
+                if (millisecondsTimeout != -1)
+                {
+                    ulong currentTime = Interop.Kernel32.GetTickCount64();
+                    ulong elapsed = currentTime - startTime;
+                        
+                    if (elapsed >= (ulong)millisecondsTimeout)
```

Comment:
We can avoid this `(ulong)` cast and `(int)` cast below by using `Environment.TickCount64` and `long` types for variables. Could you please update the code?